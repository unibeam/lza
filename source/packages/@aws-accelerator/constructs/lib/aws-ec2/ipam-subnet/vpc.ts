/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
import { IPv4CidrRange, IPv4Prefix, Pool } from 'ip-num';

import { throttlingBackOff } from '@aws-accelerator/utils/lib/throttle';
import {
  DescribeSubnetsCommand,
  DescribeSubnetsResult,
  DescribeVpcsCommand,
  EC2Client,
  Subnet,
  Tag,
  VpcCidrBlockAssociation,
} from '@aws-sdk/client-ec2';

interface IVpc {
  /**
   * The CIDRs of the VPC allocated by IPAM
   */
  readonly allocatedCidrs: IAllocatedCidr[];
  /**
   * The friendly name of the VPC
   */
  readonly name: string;
  /**
   * The ID of the VPC
   */
  readonly vpcId: string;
}

interface VpcProps {
  /**
   * The base CIDR pool of the VPC
   */
  readonly basePool: string[];
  /**
   * The ID of the VPC
   */
  readonly vpcId: string;
}

export class Vpc implements IVpc {
  public allocatedCidrs: AllocatedCidr[] = [];
  public name = '';
  public readonly vpcId: string;
  private readonly baseRanges: IPv4CidrRange[];

  constructor(props: VpcProps) {
    this.vpcId = props.vpcId;
    this.baseRanges = props.basePool.map(item => {
      return IPv4CidrRange.fromCidr(item);
    });
  }

  /**
   * Initialize the VPC object
   * @param ec2Client EC2Client
   * @returns Vpc
   */
  public async init(ec2Client: EC2Client): Promise<Vpc> {
    await this.vpcDetails(ec2Client);
    await this.subnetDetails(ec2Client);
    return this;
  }

  /**
   * Retrieve VPC details
   * @param ec2Client EC2Client
   */
  private async vpcDetails(ec2Client: EC2Client): Promise<void> {
    //
    // Get VPC details
    console.log(`Retrieving VPC details for VPC ${this.vpcId}...`);
    try {
      const vpcDetails = await throttlingBackOff(() =>
        ec2Client.send(new DescribeVpcsCommand({ VpcIds: [this.vpcId] })),
      );

      if (!vpcDetails.Vpcs || !vpcDetails.Vpcs[0].CidrBlockAssociationSet) {
        throw new Error(`Unable to retrieve CIDR block details for VPC ${this.vpcId}`);
      }

      // Determine CIDRs allocated by IPAM
      for (const baseRange of this.baseRanges) {
        for (const item of vpcDetails.Vpcs[0].CidrBlockAssociationSet ?? []) {
          // Compare VPC CIDR to base CIDR range
          const vpcRange = this.validateVpcCidr(item);
          if (vpcRange && (vpcRange.inside(baseRange) || vpcRange.isEquals(baseRange))) {
            this.setAllocatedCidr(vpcRange);
          }
        }
      }
      //
      // Get name tag
      const nameTag = vpcDetails.Vpcs[0].Tags?.filter(item => item.Key === 'Name')[0].Value;
      if (!nameTag) {
        throw new Error(`Unable to retrieve name tag for VPC ${this.vpcId}`);
      }
      //
      // Set property values
      this.name = nameTag;
    } catch (e) {
      throw new Error(`Error retrieving VPC details: ${e}`);
    }
  }

  /**
   * Validate CIDR block state and return CIDR range object
   * @param cidrItem
   * @returns IPv4CidrRange | undefined
   */
  private validateVpcCidr(cidrItem: VpcCidrBlockAssociation): IPv4CidrRange | undefined {
    if (!cidrItem.CidrBlock) {
      throw new Error(`Unable to retrieve CIDR block for VPC ${this.vpcId}`);
    }
    if (cidrItem.CidrBlockState?.State === 'associated') {
      return IPv4CidrRange.fromCidr(cidrItem.CidrBlock);
    }
    return;
  }

  /**
   * Validate API response subnet state
   * @param subnet
   */
  private validateSubnet(subnet: Subnet) {
    if (!subnet.SubnetId) {
      throw new Error(`Unable to retrieve subnet ID`);
    }
    if (!subnet.CidrBlock) {
      throw new Error(`Unable to retrieve CIDR block for subnet ${subnet.SubnetId}`);
    }
  }

  /**
   * Add subnets to the allocatedCidrs array
   * @param subnetList
   */
  private setSubnets(subnetList: DescribeSubnetsResult | undefined): void {
    for (const subnet of subnetList?.Subnets ?? []) {
      this.validateSubnet(subnet);

      // Determine if subnet is in scope of IPAM
      const subnetCidr = IPv4CidrRange.fromCidr(subnet.CidrBlock!);
      for (const vpcRange of this.allocatedCidrs) {
        if (subnetCidr.inside(vpcRange.cidr) || subnetCidr.isEquals(vpcRange.cidr)) {
          // Get name tag
          const nameTag = subnet.Tags?.filter(item => item.Key === 'Name')[0].Value;

          // Push object to array
          vpcRange.addSubnet(
            new VpcSubnet({
              allocatedCidr: subnetCidr,
              mapPublicIpOnLaunch: subnet.MapPublicIpOnLaunch ?? false,
              name: nameTag,
              subnetId: subnet.SubnetId!,
              tags: subnet.Tags,
            }),
          );
        }
      }
    }
  }

  /**
   * Retrieve subnet details for the VPC
   * @param ec2Client EC2Client
   */
  private async subnetDetails(ec2Client: EC2Client): Promise<void> {
    let nextToken: string | undefined = undefined;
    console.log(`Retrieving subnet details for VPC ${this.vpcId}...`);
    try {
      do {
        //
        // Get subnet details
        const page = await throttlingBackOff(() =>
          ec2Client.send(
            new DescribeSubnetsCommand({ Filters: [{ Name: 'vpc-id', Values: [this.vpcId] }], NextToken: nextToken }),
          ),
        );

        this.setSubnets(page);

        nextToken = page.NextToken;
      } while (nextToken);
    } catch (e) {
      throw new Error(`Error retrieving subnet details: ${e}`);
    }
  }

  /**
   * Mutator method to set an allocated CIDR
   * @param cidr IPv4CidrRange
   */
  private setAllocatedCidr(cidr: IPv4CidrRange) {
    this.allocatedCidrs.push(new AllocatedCidr(cidr));
  }

  /**
   * Accessor method to return a subnet by name
   * @param name
   * @returns VpcSubnet
   */
  public getSubnetByName(name: string): VpcSubnet {
    let subnet: VpcSubnet | undefined = undefined;

    for (const cidr of this.allocatedCidrs) {
      subnet = cidr.subnets.find(item => item.name === name);
      if (subnet) {
        break;
      }
    }
    if (!subnet) {
      throw new Error(`Subnet with Name tag "${name}" does not exist in ${this.vpcId}`);
    }
    return subnet;
  }

  /**
   * Returns the next CIDR range available in the VPC
   * @param netmaskLength bigint
   * @param vpc Vpc
   * @returns string | undefined
   */
  public allocateCidr(netmaskLength: bigint): string | undefined {
    //
    // Convert netmask to prefix
    const requestedPrefix = IPv4Prefix.fromNumber(netmaskLength);

    for (const [index, item] of this.allocatedCidrs.entries()) {
      try {
        //
        // Get next subnet CIDR range from the pool.
        // If it doesn't exist, try the next VPC CIDR range.
        const nextRange = this.getNextRangeFromPool(item.cidr, requestedPrefix);
        if (!nextRange) {
          continue;
        }
        //
        // Get next available subnet CIDR for the VPC CIDR range
        const subnetCidr = this.getNextAvailableCidr(item, nextRange);
        //
        // Determine if next CIDR is available.
        // If it is, break out of the loop.
        if (subnetCidr) {
          return subnetCidr;
        } else {
          if (index + 1 === this.allocatedCidrs.length) {
            throw new Error(
              `VPC is exhausted of address space. Cannot allocate a CIDR with /${netmaskLength.toString()} prefix.`,
            );
          }
        }
      } catch (e) {
        throw new Error(`Error while calculating next subnet CIDR: ${e}`);
      }
    }
    return;
  }

  /**
   * Creates a pool of CIDR ranges from the given VPC CIDR and returns
   * the next CIDR range from the pool, if available
   * @param vpcCidr IPv4CidrRange
   * @param requestedPrefix IPv4Prefix
   * @returns IPv4CidrRange | undefined
   */
  private getNextRangeFromPool(vpcCidr: IPv4CidrRange, requestedPrefix: IPv4Prefix): IPv4CidrRange | undefined {
    try {
      const pool = Pool.fromCidrRanges([vpcCidr]);
      return pool.getCidrRange(requestedPrefix);
    } catch (e) {
      // If the above operation fails, it's due to the requested prefix being too large.
      // Return undefined to indicate that no CIDR range could be found.
      return;
    }
  }

  /**
   * Returns the next valid CIDR range from the given VPC CIDR if it exists
   * @param vpcCidr AllocatedCidr
   * @param nextRange IPv4CidrRange
   * @returns string | undefined
   */
  private getNextAvailableCidr(vpcCidr: AllocatedCidr, nextRange: IPv4CidrRange): string | undefined {
    // Determine if the next range is overlapping.
    // If it is, try to find one that doesn't overlap.
    if (!this.isOverlapping(vpcCidr, nextRange)) {
      return nextRange.toCidrString();
    } else {
      return this.tryFindNextCidr(vpcCidr, nextRange);
    }
  }

  /**
   * Returns true if the next CIDR range is overlapping with subnets associated with the given VPC CIDR
   * @param vpcCidr AllocatedCidr
   * @param nextCidr IPv4CidrRange
   * @returns boolean
   */
  private isOverlapping(vpcCidr: AllocatedCidr, nextCidr: IPv4CidrRange): boolean {
    for (const subnet of vpcCidr.subnets) {
      if (
        subnet.allocatedCidr.contains(nextCidr) ||
        subnet.allocatedCidr.isEquals(nextCidr) ||
        nextCidr.contains(subnet.allocatedCidr)
      ) {
        return true;
      }
    }
    return false;
  }

  /**
   * Tries to find the next CIDR range that is not overlapping with the given VPC CIDR
   * @param vpcCidr AllocatedCidr
   * @param range IPv4CidrRange
   * @returns string | undefined
   */
  private tryFindNextCidr(vpcCidr: AllocatedCidr, range: IPv4CidrRange): string | undefined {
    let nextCidr = range.nextRange();
    let overlappingCidr = false;
    if (!nextCidr) {
      return;
    }

    do {
      if (!this.isOverlapping(vpcCidr, nextCidr) && vpcCidr.cidr.getLast().isGreaterThan(nextCidr.getFirst())) {
        return nextCidr.toCidrString();
      } else {
        overlappingCidr = true;
        nextCidr = nextCidr.nextRange();
        if (!nextCidr || vpcCidr.cidr.getLast().isLessThanOrEquals(nextCidr.getFirst())) {
          return;
        }
      }
    } while (overlappingCidr);
    return;
  }
}

interface IAllocatedCidr {
  /**
   * The CIDR of the allocated CIDR block
   */
  readonly cidr: IPv4CidrRange;
  /**
   * The subnets associated with the CIDR
   */
  readonly subnets: VpcSubnet[];
}

export class AllocatedCidr implements IAllocatedCidr {
  public readonly cidr: IPv4CidrRange;
  public readonly subnets: VpcSubnet[];

  constructor(cidr: IPv4CidrRange) {
    this.cidr = cidr;
    this.subnets = [];
  }

  /**
   * Mutator method to add a subnet to the allocated CIDR
   * @param subnet
   */
  public addSubnet(subnet: VpcSubnet) {
    this.subnets.push(subnet);
  }
}

interface IVpcSubnet {
  /**
   * The CIDR of the subnet allocated by IPAM
   */
  readonly allocatedCidr: IPv4CidrRange;
  /**
   * The ID of the subnet
   */
  readonly subnetId: string;
  /**
   * The MapPublicIpOnLaunch subnet attribute
   */
  readonly mapPublicIpOnLaunch?: boolean;
  /**
   * The friendly name of the subnet
   */
  readonly name?: string;
  /**
   * Tags associated with the subnet
   */
  readonly tags?: Tag[];
}

interface VpcSubnetProps {
  /**
   * The CIDR of the subnet allocated by IPAM
   */
  readonly allocatedCidr: IPv4CidrRange;
  /**
   * The ID of the subnet
   */
  readonly subnetId: string;
  /**
   * The MapPublicIpOnLaunch subnet attribute
   */
  readonly mapPublicIpOnLaunch?: boolean;
  /**
   * The friendly name of the subnet
   */
  readonly name?: string;
  /**
   * Tags associated with the subnet
   */
  readonly tags?: Tag[];
}

export class VpcSubnet implements IVpcSubnet {
  public readonly allocatedCidr: IPv4CidrRange;
  public readonly subnetId: string;
  public readonly mapPublicIpOnLaunch: boolean | undefined;
  public readonly name: string | undefined;
  public readonly tags: Tag[] | undefined;

  constructor(props: VpcSubnetProps) {
    this.allocatedCidr = props.allocatedCidr;
    this.mapPublicIpOnLaunch = props.mapPublicIpOnLaunch;
    this.name = props.name;
    this.subnetId = props.subnetId;
    this.tags = props.tags;
  }
}

export async function vpcInit(props: VpcProps, ec2Client: EC2Client): Promise<Vpc> {
  // Set static properties
  const vpc = new Vpc(props);

  // Set VPC and subnet properties
  await vpc.init(ec2Client);
  return vpc;
}
