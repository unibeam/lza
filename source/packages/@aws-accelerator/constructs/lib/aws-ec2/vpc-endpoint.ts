/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

import { ISecurityGroup } from './vpc';

export interface IVpcEndpoint extends cdk.IResource {
  readonly vpcEndpointId: string;
  readonly service: string;
  readonly vpcId: string;
  readonly dnsName?: string;
  readonly hostedZoneId?: string;

  createEndpointRoute: (id: string, routeTableId: string, destination?: string, ipv6Destination?: string) => void;
}

export enum VpcEndpointType {
  INTERFACE = 'Interface',
  GATEWAY = 'Gateway',
  GWLB = 'GatewayLoadBalancer',
}

export interface VpcEndpointProps {
  readonly vpcEndpointType: VpcEndpointType;
  readonly service: string;
  readonly vpcId: string;
  readonly subnets?: string[];
  readonly securityGroups?: ISecurityGroup[];
  readonly privateDnsEnabled?: boolean;
  readonly policyDocument?: cdk.aws_iam.PolicyDocument;
  readonly routeTables?: string[];
  readonly partition?: string;
  readonly serviceName?: string;
}

abstract class VpcEndpointBase extends cdk.Resource implements IVpcEndpoint {
  public abstract readonly vpcEndpointId: string;
  public abstract readonly vpcId: string;
  public abstract readonly service: string;
  public abstract readonly dnsName?: string;
  public abstract readonly hostedZoneId?: string;

  public createEndpointRoute(id: string, routeTableId: string, destination?: string, ipv6Destination?: string): void {
    new cdk.aws_ec2.CfnRoute(this, id, {
      destinationCidrBlock: destination,
      destinationIpv6CidrBlock: ipv6Destination,
      routeTableId,
      vpcEndpointId: this.vpcEndpointId,
    });
  }
}

interface VpcEndpointAttributes {
  vpcId: string;
  vpcEndpointId: string;
  service: string;
}

export class VpcEndpoint extends VpcEndpointBase {
  public readonly vpcEndpointId: string;
  public readonly vpcId: string;
  public readonly service: string;
  public readonly dnsName?: string;
  public readonly hostedZoneId?: string;

  static fromAttributes(scope: Construct, id: string, attrs: VpcEndpointAttributes): IVpcEndpoint {
    class Import extends VpcEndpointBase {
      public readonly vpcEndpointId = attrs.vpcEndpointId;
      public readonly vpcId = attrs.vpcId;
      public readonly service = attrs.service;
      public readonly dnsName?: string;
      public readonly hostedZoneId?: string;
      constructor(scope: Construct, id: string) {
        super(scope, id);
      }
    }
    return new Import(scope, id);
  }

  constructor(scope: Construct, id: string, props: VpcEndpointProps) {
    super(scope, id);

    this.service = props.service;
    this.vpcId = props.vpcId;

    // Add constant for sagemaker conditionals
    const sagemakerArray = ['notebook', 'studio'];

    if (props.vpcEndpointType === VpcEndpointType.INTERFACE) {
      let serviceName = `com.amazonaws.${cdk.Stack.of(this).region}.${props.service}`;
      if (sagemakerArray.includes(this.service)) {
        serviceName = `aws.sagemaker.${cdk.Stack.of(this).region}.${props.service}`;
      }
      if (this.service === 's3-global.accesspoint') {
        serviceName = `com.amazonaws.${props.service}`;
      }
      // Add the ability against China region to override serviceName due to the prefix of
      // serviceName is inconsistent (com.amazonaws vs cn.com.amazonaws) for VPC interface
      // endpoints in that region.
      if (props.serviceName) {
        serviceName = props.serviceName;
      }
      const resource = new cdk.aws_ec2.CfnVPCEndpoint(this, 'Resource', {
        serviceName,
        vpcEndpointType: props.vpcEndpointType,
        vpcId: this.vpcId,
        subnetIds: props.subnets,
        securityGroupIds: props.securityGroups?.map(item => item.securityGroupId),
        privateDnsEnabled: props.privateDnsEnabled,
        policyDocument: props.policyDocument,
      });
      this.vpcEndpointId = resource.ref;

      let dnsEntriesIndex = 0;
      if (sagemakerArray.includes(this.service)) {
        dnsEntriesIndex = 4;
      }

      this.dnsName = cdk.Fn.select(1, cdk.Fn.split(':', cdk.Fn.select(dnsEntriesIndex, resource.attrDnsEntries)));
      this.hostedZoneId = cdk.Fn.select(0, cdk.Fn.split(':', cdk.Fn.select(dnsEntriesIndex, resource.attrDnsEntries)));
      return;
    } else if (props.vpcEndpointType === VpcEndpointType.GATEWAY) {
      let serviceName = new cdk.aws_ec2.GatewayVpcEndpointAwsService(props.service).name;
      if (props.serviceName) {
        serviceName = props.serviceName;
      }
      const resource = new cdk.aws_ec2.CfnVPCEndpoint(this, 'Resource', {
        serviceName,
        vpcId: this.vpcId,
        routeTableIds: props.routeTables,
        policyDocument: props.policyDocument,
      });
      this.vpcEndpointId = resource.ref;
      return;
    } else {
      const servicePrefix = props.partition === 'aws-cn' ? 'cn.com.amazonaws' : 'com.amazonaws';
      const serviceName = `${servicePrefix}.vpce.${cdk.Stack.of(this).region}.${props.service}`;

      const resource = new cdk.aws_ec2.CfnVPCEndpoint(this, 'Resource', {
        serviceName,
        vpcEndpointType: props.vpcEndpointType,
        vpcId: this.vpcId,
        subnetIds: props.subnets,
      });
      this.vpcEndpointId = resource.ref;
      return;
    }
  }
}
