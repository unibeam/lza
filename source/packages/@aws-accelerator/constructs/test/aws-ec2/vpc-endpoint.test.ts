/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */

import * as cdk from 'aws-cdk-lib';
import { SecurityGroup } from '../../lib/aws-ec2/vpc';
import { VpcEndpoint, VpcEndpointType } from '../../lib/aws-ec2/vpc-endpoint';
import { snapShotTest } from '../snapshot-test';
import { describe, it, expect } from '@jest/globals';

const testNamePrefix = 'Construct(VpcEndpoint): ';

//Initialize stack for snapshot test and resource configuration test
const stack = new cdk.Stack();

const securityGroup = new SecurityGroup(stack, 'TestSecurityGroup`', {
  securityGroupName: 'TestSecurityGroup',
  description: `AWS Private Endpoint Zone`,
  vpcId: 'Test',
});

/**
 * VpcEndpoint construct test
 */
describe('VpcEndpoint', () => {
  it('vpc gateway end point type test', () => {
    const initialStack = new VpcEndpoint(stack, 'VpcEndpoint', {
      vpcId: 'Test',
      vpcEndpointType: VpcEndpointType.GATEWAY,
      service: 'service',
      subnets: ['Test1', 'Test2'],
      securityGroups: [securityGroup],
      privateDnsEnabled: true,
      policyDocument: new cdk.aws_iam.PolicyDocument({
        statements: [
          new cdk.aws_iam.PolicyStatement({
            sid: 'AccessToTrustedPrincipalsAndResources',
            actions: ['*'],
            effect: cdk.aws_iam.Effect.ALLOW,
            resources: ['*'],
            principals: [new cdk.aws_iam.AnyPrincipal()],
            conditions: {
              StringEquals: {
                'aws:PrincipalOrgID': ['organizationId'],
              },
            },
          }),
        ],
      }),
      routeTables: ['Test1', 'Test2'],
    });
    expect(typeof initialStack.createEndpointRoute('id', 'routeTableId', '10.100.0.0/16')).toBe('undefined');
  });
  it('vpc interface end point type test with sagemaker', () => {
    new VpcEndpoint(stack, 'VpcEndpointInterfaceSagemaker', {
      vpcId: 'Test',
      vpcEndpointType: VpcEndpointType.INTERFACE,
      service: 'notebook',
      subnets: ['Test1', 'Test2'],
      securityGroups: [securityGroup],
      privateDnsEnabled: true,
      policyDocument: new cdk.aws_iam.PolicyDocument({
        statements: [
          new cdk.aws_iam.PolicyStatement({
            sid: 'AccessToTrustedPrincipalsAndResources',
            actions: ['*'],
            effect: cdk.aws_iam.Effect.ALLOW,
            resources: ['*'],
            principals: [new cdk.aws_iam.AnyPrincipal()],
            conditions: {
              StringEquals: {
                'aws:PrincipalOrgID': ['organizationId'],
              },
            },
          }),
        ],
      }),
      routeTables: ['Test1', 'Test2'],
    });
  });
  it('vpc interface end point type test with s3 global access', () => {
    new VpcEndpoint(stack, 'VpcEndpointInterfaceS3', {
      vpcId: 'Test',
      vpcEndpointType: VpcEndpointType.INTERFACE,
      service: 's3-global.accesspoint',
      subnets: ['Test1', 'Test2'],
      securityGroups: [securityGroup],
      privateDnsEnabled: true,
      policyDocument: new cdk.aws_iam.PolicyDocument({
        statements: [
          new cdk.aws_iam.PolicyStatement({
            sid: 'AccessToTrustedPrincipalsAndResources',
            actions: ['*'],
            effect: cdk.aws_iam.Effect.ALLOW,
            resources: ['*'],
            principals: [new cdk.aws_iam.AnyPrincipal()],
            conditions: {
              StringEquals: {
                'aws:PrincipalOrgID': ['organizationId'],
              },
            },
          }),
        ],
      }),
      routeTables: ['Test1', 'Test2'],
    });
  });
  it('vpc interface end point type test with serviceName', () => {
    new VpcEndpoint(stack, 'VpcEndpointInterfaceEc2', {
      vpcId: 'Test',
      vpcEndpointType: VpcEndpointType.INTERFACE,
      serviceName: 'ec2',
      service: 'ec2',
      subnets: ['Test1', 'Test2'],
      securityGroups: [securityGroup],
      privateDnsEnabled: true,
      policyDocument: new cdk.aws_iam.PolicyDocument({
        statements: [
          new cdk.aws_iam.PolicyStatement({
            sid: 'AccessToTrustedPrincipalsAndResources',
            actions: ['*'],
            effect: cdk.aws_iam.Effect.ALLOW,
            resources: ['*'],
            principals: [new cdk.aws_iam.AnyPrincipal()],
            conditions: {
              StringEquals: {
                'aws:PrincipalOrgID': ['organizationId'],
              },
            },
          }),
        ],
      }),
      routeTables: ['Test1', 'Test2'],
    });
  });
  it('vpc interface end point type test with gwlb', () => {
    new VpcEndpoint(stack, 'VpcEndpointInterfaceGwlb', {
      vpcId: 'Test',
      vpcEndpointType: VpcEndpointType.GWLB,
      serviceName: 'ec2',
      service: 'ec2',
      subnets: ['Test1', 'Test2'],
      securityGroups: [securityGroup],
      privateDnsEnabled: true,
      policyDocument: new cdk.aws_iam.PolicyDocument({
        statements: [
          new cdk.aws_iam.PolicyStatement({
            sid: 'AccessToTrustedPrincipalsAndResources',
            actions: ['*'],
            effect: cdk.aws_iam.Effect.ALLOW,
            resources: ['*'],
            principals: [new cdk.aws_iam.AnyPrincipal()],
            conditions: {
              StringEquals: {
                'aws:PrincipalOrgID': ['organizationId'],
              },
            },
          }),
        ],
      }),
      routeTables: ['Test1', 'Test2'],
    });
  });

  it('vpc gateway end point for s3', () => {
    VpcEndpoint.fromAttributes(stack, 'ImportedVpcEndpointS3', {
      service: 's3',
      vpcEndpointId: 'importedEndpointId',
      vpcId: 'Test',
    });
  });
  it('serviceName override for gatewayEndpoint', () => {
    const checkVpcEndpointGatewayEndpointServiceNameStack = new cdk.Stack();
    new VpcEndpoint(checkVpcEndpointGatewayEndpointServiceNameStack, 'VpcEndpointGatewayEndpointServiceName', {
      vpcId: 'Test',
      vpcEndpointType: VpcEndpointType.GATEWAY,
      service: 'service',
      serviceName: 'testGatewayEndpointServiceName',
      subnets: ['Test1', 'Test2'],
      securityGroups: [securityGroup],
      privateDnsEnabled: true,
      policyDocument: new cdk.aws_iam.PolicyDocument({
        statements: [
          new cdk.aws_iam.PolicyStatement({
            sid: 'AccessToTrustedPrincipalsAndResources',
            actions: ['*'],
            effect: cdk.aws_iam.Effect.ALLOW,
            resources: ['*'],
            principals: [new cdk.aws_iam.AnyPrincipal()],
            conditions: {
              StringEquals: {
                'aws:PrincipalOrgID': ['organizationId'],
              },
            },
          }),
        ],
      }),
      routeTables: ['Test1', 'Test2'],
    });
    const checkVpcEndpointGatewayEndpointServiceNameTemplate = cdk.assertions.Template.fromStack(
      checkVpcEndpointGatewayEndpointServiceNameStack,
    );
    checkVpcEndpointGatewayEndpointServiceNameTemplate.hasResourceProperties('AWS::EC2::VPCEndpoint', {
      ServiceName: cdk.assertions.Match.exact('testGatewayEndpointServiceName'),
    });
  });

  snapShotTest(testNamePrefix, stack);
});
