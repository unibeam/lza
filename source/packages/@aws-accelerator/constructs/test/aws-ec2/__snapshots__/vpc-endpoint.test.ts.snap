// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VpcEndpoint Construct(VpcEndpoint):  Snapshot Test 1`] = `
{
  "Resources": {
    "TestSecurityGroupDA4B5F83": {
      "Properties": {
        "GroupDescription": "AWS Private Endpoint Zone",
        "GroupName": "TestSecurityGroup",
        "Tags": [
          {
            "Key": "Name",
            "Value": "TestSecurityGroup",
          },
        ],
        "VpcId": "Test",
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "VpcEndpoint80208C18": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "*",
              "Condition": {
                "StringEquals": {
                  "aws:PrincipalOrgID": [
                    "organizationId",
                  ],
                },
              },
              "Effect": "Allow",
              "Principal": {
                "AWS": "*",
              },
              "Resource": "*",
              "Sid": "AccessToTrustedPrincipalsAndResources",
            },
          ],
          "Version": "2012-10-17",
        },
        "RouteTableIds": [
          "Test1",
          "Test2",
        ],
        "ServiceName": {
          "Fn::Join": [
            "",
            [
              "com.amazonaws.",
              {
                "Ref": "AWS::Region",
              },
              ".service",
            ],
          ],
        },
        "VpcId": "Test",
      },
      "Type": "AWS::EC2::VPCEndpoint",
    },
    "VpcEndpointInterfaceEc21C025642": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "*",
              "Condition": {
                "StringEquals": {
                  "aws:PrincipalOrgID": [
                    "organizationId",
                  ],
                },
              },
              "Effect": "Allow",
              "Principal": {
                "AWS": "*",
              },
              "Resource": "*",
              "Sid": "AccessToTrustedPrincipalsAndResources",
            },
          ],
          "Version": "2012-10-17",
        },
        "PrivateDnsEnabled": true,
        "SecurityGroupIds": [
          {
            "Ref": "TestSecurityGroupDA4B5F83",
          },
        ],
        "ServiceName": "ec2",
        "SubnetIds": [
          "Test1",
          "Test2",
        ],
        "VpcEndpointType": "Interface",
        "VpcId": "Test",
      },
      "Type": "AWS::EC2::VPCEndpoint",
    },
    "VpcEndpointInterfaceGwlb3F654CE2": {
      "Properties": {
        "ServiceName": {
          "Fn::Join": [
            "",
            [
              "com.amazonaws.vpce.",
              {
                "Ref": "AWS::Region",
              },
              ".ec2",
            ],
          ],
        },
        "SubnetIds": [
          "Test1",
          "Test2",
        ],
        "VpcEndpointType": "GatewayLoadBalancer",
        "VpcId": "Test",
      },
      "Type": "AWS::EC2::VPCEndpoint",
    },
    "VpcEndpointInterfaceS343EF5B95": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "*",
              "Condition": {
                "StringEquals": {
                  "aws:PrincipalOrgID": [
                    "organizationId",
                  ],
                },
              },
              "Effect": "Allow",
              "Principal": {
                "AWS": "*",
              },
              "Resource": "*",
              "Sid": "AccessToTrustedPrincipalsAndResources",
            },
          ],
          "Version": "2012-10-17",
        },
        "PrivateDnsEnabled": true,
        "SecurityGroupIds": [
          {
            "Ref": "TestSecurityGroupDA4B5F83",
          },
        ],
        "ServiceName": "com.amazonaws.s3-global.accesspoint",
        "SubnetIds": [
          "Test1",
          "Test2",
        ],
        "VpcEndpointType": "Interface",
        "VpcId": "Test",
      },
      "Type": "AWS::EC2::VPCEndpoint",
    },
    "VpcEndpointInterfaceSagemaker70CE2242": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "*",
              "Condition": {
                "StringEquals": {
                  "aws:PrincipalOrgID": [
                    "organizationId",
                  ],
                },
              },
              "Effect": "Allow",
              "Principal": {
                "AWS": "*",
              },
              "Resource": "*",
              "Sid": "AccessToTrustedPrincipalsAndResources",
            },
          ],
          "Version": "2012-10-17",
        },
        "PrivateDnsEnabled": true,
        "SecurityGroupIds": [
          {
            "Ref": "TestSecurityGroupDA4B5F83",
          },
        ],
        "ServiceName": {
          "Fn::Join": [
            "",
            [
              "aws.sagemaker.",
              {
                "Ref": "AWS::Region",
              },
              ".notebook",
            ],
          ],
        },
        "SubnetIds": [
          "Test1",
          "Test2",
        ],
        "VpcEndpointType": "Interface",
        "VpcId": "Test",
      },
      "Type": "AWS::EC2::VPCEndpoint",
    },
    "VpcEndpointid71FBEB7E": {
      "Properties": {
        "DestinationCidrBlock": "10.100.0.0/16",
        "RouteTableId": "routeTableId",
        "VpcEndpointId": {
          "Ref": "VpcEndpoint80208C18",
        },
      },
      "Type": "AWS::EC2::Route",
    },
  },
}
`;
